contracts/gov/src/contract.rs

```diff
@@ -420,225 +423,210 @@ pub fn end_poll<S: Storage, A: Api, Q: Querier>(
+    ]))
 }
 
 /*
- * Execute a msg of passed poll.
+ * Execute a msgs of passed poll as one submsg to catch failures
  */
-pub fn execute_poll<S: Storage, A: Api, Q: Querier>(
-    deps: &mut Extern<S, A, Q>,
-    env: Env,
-    poll_id: u64,
-) -> HandleResult {
-    let config: Config = config_read(&deps.storage).load()?;
-    let mut a_poll: Poll = poll_store(&mut deps.storage).load(&poll_id.to_be_bytes())?;
+pub fn execute_poll(deps: DepsMut, env: Env, poll_id: u64) -> Result<Response, ContractError> {
+    let config: Config = config_read(deps.storage).load()?;
+    let a_poll: Poll = poll_store(deps.storage).load(&poll_id.to_be_bytes())?;
 
     if a_poll.status != PollStatus::Passed {
-        return Err(StdError::generic_err("Poll is not in passed status"));
+        return Err(ContractError::PollNotPassed {});
     }
 
     if a_poll.end_height + config.timelock_period > env.block.height {
-        return Err(StdError::generic_err("Timelock period has not expired"));
+        return Err(ContractError::TimelockNotExpired {});
     }
 
-    poll_indexer_store(&mut deps.storage, &PollStatus::Passed).remove(&poll_id.to_be_bytes());
-    poll_indexer_store(&mut deps.storage, &PollStatus::Executed)
-        .save(&poll_id.to_be_bytes(), &true)?;
+    store_tmp_poll_id(deps.storage, a_poll.id)?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_error(
+        CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: env.contract.address.to_string(),
+            msg: to_binary(&ExecuteMsg::ExecutePollMsgs { poll_id })?,
+            funds: vec![],
+        }),
+        POLL_EXECUTE_REPLY_ID,
+    )))
+}
+
+/*
+ * Execute a msgs of a poll
+ */
+pub fn execute_poll_messages(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    poll_id: u64,
+) -> Result<Response, ContractError> {
+    if env.contract.address != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut a_poll: Poll = poll_store(deps.storage).load(&poll_id.to_be_bytes())?;
+
+    poll_indexer_store(deps.storage, &PollStatus::Passed).remove(&poll_id.to_be_bytes());
+    poll_indexer_store(deps.storage, &PollStatus::Executed).save(&poll_id.to_be_bytes(), &true)?;
 
     a_poll.status = PollStatus::Executed;
-    poll_store(&mut deps.storage).save(&poll_id.to_be_bytes(), &a_poll)?;
+    poll_store(deps.storage).save(&poll_id.to_be_bytes(), &a_poll)?;
 
     let mut messages: Vec<CosmosMsg> = vec![];
     if let Some(all_msgs) = a_poll.execute_data {
         let mut msgs = all_msgs;
         msgs.sort();
         for msg in msgs {
             messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
-                contract_addr: deps.api.human_address(&msg.contract)?,
+                contract_addr: deps.api.addr_humanize(&msg.contract)?.to_string(),
                 msg: msg.msg,
-                send: vec![],
-            }))
+                funds: vec![],
+            }));
         }
     } else {
```

Comment:
@ezaanm since we introduced this function for expiration, in case of a poll with an  empty/`None`  `execution_message`, if anybody sends execute message for that specific poll, it leads to poll expiration. It is better to handle the poll in this case as 'executed`.